proposals/XXXX-extend-property-wrappers-to-function-and-closure-parameters.md

```diff
@@ -0,0 +1,426 @@
+# Extend Property Wrappers to Function and Closure Parameters
+
+* Proposal: [SE-NNNN](NNNN-extend-property-wrappers-to-function-and-closure-parameters.md)
+* Authors: [Holly Borla](https://github.com/hborla), [Filip Sakel](https://github.com/filip-sakel)
+* Review Manager: TBD
+* Status: **Awaiting implementation**
+* Implementation: [apple/swift#34272](https://github.com/apple/swift/pull/34272)
+
+
+## Introduction
+
+Property Wrappers were [introduced in Swift 5.1](https://github.com/apple/swift-evolution/blob/main/proposals/0258-property-wrappers.md), and have since become a popular feature abstracting away common accessor patterns for properties. Currently, applying a property wrapper is solely permitted at local and type context. However, with increasing adoption, demand for extending _where_ property wrappers can be applied has emerged. This proposal aims to extend property wrappers to function and closure parameters.
+
+
+## Motivation
+
+Property wrappers have undoubtably been very successful. Applying a property wrapper to a property is enabled by an incredibly lightweight and expressive syntax. Therefore, library authors can expose complex behavior through easily understandable property-wrapper types in an efficient manner. For instance, frameworks such as [SwiftUI](https://developer.apple.com/documentation/swiftui/) and [Combine](https://developer.apple.com/documentation/combine) introduce property wrappers such as [`State`](https://developer.apple.com/documentation/swiftui/state), [`Binding`](https://developer.apple.com/documentation/swiftui/binding) and [`Published`](https://developer.apple.com/documentation/combine/published) respectively, to expose elaborate behavior through a succinct interface, helping craft expressive yet simple APIs. However, property wrappers are only applicable to local variables and type properties, shattering the illusion that they helped realize in the first place when working with parameters.
+
+### Memberwise initialization
+
+Currently, property-wrapped struct properties interact with function parameters, through the struct's synthesized memberwise initializer. However, property-wrapper attributes are _not_ supported on function parameters. This leads to really complicated and nuanced rules for which type, between the wrapped type and the backing property-wrapper type, the memberwise initializer accepts. Namely, the compiler will choose the wrapped-value type to offer a convenience to the call-site, when the property wrapper supports initalization through the suitable initializer: `init(wrappedValue:)`, as seen here:
```

Comment:
Here, and in some other places in the text, you use "call-site" hyphenated, yet in some other places it's unhyphenated. So, I'm asking you which one do you think is more correct, so that we can use one of the two throughout the proposal. (I assume it's "call-site", since you used it in Memberwise Initializers, which you wrote more recently; but I wanted to ask you just in case)

Reply:
I've seen them used interchangeably and I think I'm indifferent; both seem correct to me. Let's use "call-site" since that's apparently what I wrote everywhere else haha (I think it was out of habit, rather than an intentional spelling)